File: julia/12_image-compression.jl
#!/usr/bin/env julia
"""
Compress and reconstruct grayscale images using SVD
"""

using Images, ImageIO
using LinearAlgebra
using Plots

function main()
    path = "../data/sqai-square-gray-rgb150ppi.jpg"  # path to input image

    # load image and convert to grayscale
    image = load(path)::Matrix{RGB{N0f8}}
    A = channelview(Gray.(image)) .* 255.0
    h, w = size(A)
    println("image size; $h $w\n")
    plt = heatmap(A; c=:greys, clim=(0, 255), aspect_ratio=:equal, yflip=true,
        axis=nothing, title="original image")
    display(plt)
    readline()

    # SVD
    F = svd(A; full=false)
    U, S, V = F.U, F.S, F.V
    plt = plot(S; yscale=:log10, xlabel="index", ylabel="λ_i", title="singular values")
    display(plt)
    readline()

    # image reconstruction with different ranks
    ranks = [1, 2, 4, 8, 16, 32, 64, 128, 256]
    for r in ranks
        rr = min(r, length(S))
        Ar = @view(U[:, 1:rr]) * Diagonal(S[1:rr]) * transpose(@view(V[:, 1:rr]))
        plt = heatmap(Ar; c=:greys, clim=(0, 255), aspect_ratio=:equal, yflip=true,
            axis=nothing, title="reconstructed image (rank $(rr))")
        display(plt)
        readline()
    end
end

main()

File: python/12_image-compression.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Compress and reconstruct grayscale images using SVD
"""

import numpy as np
from PIL import Image
import matplotlib.pyplot as plt


def main():
    path = "../data/sqai-square-gray-rgb150ppi.jpg"  # path to input image

    """load image and convert to grayscale"""
    image = Image.open(path).convert("L")
    array = np.asarray(image, dtype=np.float64)
    h, w = array.shape
    print(f"image size; {h} {w}\n")
    plt.figure()
    plt.imshow(array, cmap="gray", vmin=0, vmax=255)
    plt.title("original image")
    plt.axis("off")
    plt.show()

    """SVD"""
    U, S, Vt = np.linalg.svd(array, full_matrices=False)
    plt.figure()
    plt.semilogy(S)
    plt.title("singular values")
    plt.xlabel("index")
    plt.ylabel("λ_i")
    plt.show()

    """image reconstruction with different ranks"""
    ranks = [1, 2, 4, 8, 16, 32, 64, 128, 256]
    for r in ranks:
        rr = min(r, len(S))
        Ur = U[:, :rr]
        Sr = np.diag(S[:rr])
        Vtr = Vt[:rr, :]
        Ar = Ur @ Sr @ Vtr
        plt.figure()
        plt.imshow(Ar, cmap="gray", vmin=0, vmax=255)
        plt.title(f"reconstructed image (rank {r})")
        plt.axis("off")
        plt.show()


if __name__ == "__main__":
    main()
